core/src/main/scala/cats/FunctorEmpty.scala

```diff
@@ -0,0 +1,28 @@
+package cats
+
+import simulacrum.typeclass
+
+/**
+ * `FunctorEmpty[F]` allows you to `map` and filter out elements simultaneously.
+ */
+@typeclass
+trait FunctorEmpty[F[_]] extends Serializable {
+  def functor: Functor[F]
+
+  def mapFilter[A, B](fa: F[A])(f: A => Option[B]): F[B]
+
+  def collect[A, B](fa: F[A])(f: PartialFunction[A, B]): F[B] =
+    mapFilter(fa)(f.lift)
+
+  def flattenOption[A](fa: F[Option[A]]): F[A] =
+    mapFilter(fa)(identity)
+
+  def filter[A](fa: F[A])(f: A => Boolean): F[A] =
+    mapFilter(fa)(a => if (f(a)) Some(a) else None)
+}
+
+object FunctorEmpty {
+  implicit def catsFunctorForFunctorEmpty[F[_]](fe: FunctorEmpty[F]): Functor[F] =
+    fe.functor
```

Comment:
This function here allows us to call things like `FunctorEmpty[F].map`, do you think this is worthwhile? I feel like this is an easy change that doesn't cause any ambiguities and let's us atleast have some of the niceties that we lose when doing composition instead of inheritance :)

Reply:
I am not sure if it's needed either. 
What would be the use case of it? 
If user `import FunctorEmpty._` and all of a sudden whenever there is a `FunctorEmpty` in scope there is also a `Functor` in scope, then will there be ambiguities with other possible `Functor` instances?
Should user be just writing the following instead?  
```def foo[F[_]: FunctorEmpty : Functor]: ....```

Reply:
There won't be ambiguities, because only an explicit instance will be converted, notice it is `(fe: FunctorEmpty[F]): Functor[F]` and not `(implicit fe: FunctorEmpty[F]): Functor[F]`.

The only real benefit is that when you have a method like this:
```scala
def foo[F[_]](fa: F[Int])(implicit F: FunctorEmpty[F]): F[String] =
  F.map(fa)(_.toString) 
```
You'll be able to call `map` et al. directly on it instead of doing `F.functor.map`. So the benefit is rather small, but at the same time it doesn't cause ambiguities. Also note that `import FunctorEmpty._` won't actually do anything as this instance will always be in scope whenever `FunctorEmpty` is in scope :)

Reply:
ah, sorry I missed that `fe` isn't implicit. As you said, the benefit is a bit small, especially since you usually don't need to write `implicit F: FunctorEmpty[F]` at the first place, you could just use context bound, so not much typing to be saved. 
I'd prefer encouraging users to the more conventional type class syntax enabled style. 
```scala
def foo[F[_]: FunctorEmpty : Functor](fa: F[Int]): F[String] = {
   fa.map(_.toString).filter(_.length > 2)
}
```

Reply:
Sounds good to me :)